File: examples/workflows/cpp/casbot_cpp_test/src/t01_get_state.cpp
// T01: 查询机器人运控状态
// 对应接口文档 5.3 获取机器人状态
#include <rclcpp/rclcpp.hpp>
#include <thread>
#include <crb_ros_msg/srv/get_robot_state.hpp>
#include <std_msgs/msg/string.hpp>

static const std::map<int, std::string> STATE_MAP = {
    {0, "UNDEFINED"},
    {1, "DAMPING"},
    {2, "READY"},
    {3, "STAND"},
    {4, "WALKING"},
    {5, "RUNNING"},
    {255, "EMERGENCY_STOP"},
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = rclcpp::Node::make_shared("t01_get_state");

    // --- 1. 通过 Service 查询状态 ---
    auto client = node->create_client<crb_ros_msg::srv::GetRobotState>("get_robot_state_srv_hl");
    RCLCPP_INFO(node->get_logger(), "[T01] 等待 /get_robot_state 服务...");
    if (!client->wait_for_service(std::chrono::seconds(5))) {
        RCLCPP_ERROR(node->get_logger(), "[T01] 服务不可用，超时退出");
        rclcpp::shutdown();
        return 1;
    }

    auto req = std::make_shared<crb_ros_msg::srv::GetRobotState::Request>();
    req->start = true;
    auto future = client->async_send_request(req);

    if (rclcpp::spin_until_future_complete(node, future) == rclcpp::FutureReturnCode::SUCCESS) {
        int state = static_cast<int>(future.get()->state);
        auto it = STATE_MAP.find(state);
        std::string name = (it != STATE_MAP.end()) ? it->second : "UNKNOWN";
        RCLCPP_INFO(node->get_logger(),
            "[T01] GetRobotState => state=%d (%s)", state, name.c_str());
    } else {
        RCLCPP_ERROR(node->get_logger(), "[T01] Service 调用失败");
    }

    // --- 2. 通过 Topic 订阅状态字符串 ---
    RCLCPP_INFO(node->get_logger(), "[T01] 订阅 /motion/status (3 秒)...");
    std::string last_status;
    auto sub = node->create_subscription<std_msgs::msg::String>(
        "/motion/status", 10,
        [&last_status, &node](const std_msgs::msg::String::SharedPtr msg) {
            if (last_status != msg->data) {
                last_status = msg->data;
                RCLCPP_INFO(node->get_logger(), "[T01] /motion/status: %s", msg->data.c_str());
            }
        });

    auto deadline = node->now() + rclcpp::Duration(3, 0);
    while (rclcpp::ok() && node->now() < deadline) {
        rclcpp::spin_some(node);
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }

    RCLCPP_INFO(node->get_logger(), "[T01] 完成");
    rclcpp::shutdown();
    return 0;
}

File: examples/workflows/python/casbot_py_test/t01_get_state.py
#!/usr/bin/env python3
"""T01: 查询机器人运控状态 — 对应接口文档 5.3"""
import rclpy
from rclpy.node import Node
from crb_ros_msg.srv import GetRobotState
from std_msgs.msg import String

STATE_MAP = {
    0: 'UNDEFINED', 1: 'DAMPING', 2: 'READY',
    3: 'STAND',     4: 'WALKING', 5: 'RUNNING',
    255: 'EMERGENCY_STOP',
}

def main():
    rclpy.init()
    node = Node('t01_get_state_py')

    # --- 1. Service 查询状态 ---
    cli = node.create_client(GetRobotState, 'get_robot_state_srv_hl')
    node.get_logger().info('[T01] 等待 /get_robot_state 服务...')
    if not cli.wait_for_service(timeout_sec=5.0):
        node.get_logger().error('[T01] 服务不可用，超时退出')
        rclpy.shutdown(); return

    req = GetRobotState.Request()
    req.start = True
    future = cli.call_async(req)
    rclpy.spin_until_future_complete(node, future, timeout_sec=5.0)
    resp = future.result()
    if resp:
        state = resp.state
        name = STATE_MAP.get(state, 'UNKNOWN')
        node.get_logger().info(f'[T01] GetRobotState => state={state} ({name})')
    else:
        node.get_logger().error('[T01] Service 调用失败')

    # --- 2. Topic 订阅状态字符串 ---
    node.get_logger().info('[T01] 订阅 /motion/status (3 秒)...')
    last = ['']
    def status_cb(msg: String):
        if msg.data != last[0]:
            last[0] = msg.data
            node.get_logger().info(f'[T01] /motion/status: {msg.data}')
    node.create_subscription(String, '/motion/status', status_cb, 10)

    import time; deadline = time.time() + 3.0
    while time.time() < deadline and rclpy.ok():
        rclpy.spin_once(node, timeout_sec=0.1)

    node.get_logger().info('[T01] 完成')
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
